components/support/migration/src/main/java/mozilla/components/support/migration/AddonMigration.kt

```diff
@@ -0,0 +1,135 @@
+    /**
+     * Failure variants of an add-on migration.
+     */
+    sealed class Failure : AddonMigrationResult() {
+        /**
+         * Failed to query installed add-ons via the [Engine].
+         */
+        internal data class FailedToQueryInstalledAddons(val throwable: Throwable) : Failure() {
+            override fun toString(): String {
+                return "Failed to query installed add-ons: ${throwable::class}"
+            }
+        }
+
+        /**
+         * Failed to migrate some or all of the installed add-ons.
+         */
+        internal data class FailedToMigrateAddons(
+            val migratedAddons: List<WebExtension>,
+            val failedAddons: Map<WebExtension, Throwable>
+        ) : Failure() {
+            override fun toString(): String {
+                return "Failed to migrate ${failedAddons.size} of ${failedAddons.size + migratedAddons.size} add-ons."
+            }
+        }
+    }
+}
+
+/**
+ * Helper for migrating add-ons.
+ */
+internal object AddonMigration {
+
+    /**
+     * Performs a migration of all installed add-ons. The only migration step involved
+     * is to make sure we disable all currently unsupported add-ons.
+     */
+    @SuppressWarnings("TooGenericExceptionCaught")
+    suspend fun migrate(engine: Engine): Result<AddonMigrationResult> {
+        try {
+            val installedAddons = getInstalledAddons(engine)
+            if (installedAddons.isEmpty()) {
+                return Result.Success(AddonMigrationResult.Success.NoAddons)
+            }
+
+            val migratedAddons = mutableListOf<WebExtension>()
+            val failures = mutableMapOf<WebExtension, Exception>()
+            installedAddons.forEach { addon ->
+                try {
+                    migratedAddons += migrateAddon(engine, addon)
+                } catch (e: Exception) {
+                    failures[addon] = e
+                }
+            }
+
+            return if (failures.isEmpty()) {
+                Result.Success(AddonMigrationResult.Success.AddonsMigrated(migratedAddons))
+            } else {
+                val failure = AddonMigrationResult.Failure.FailedToMigrateAddons(migratedAddons, failures)
+                Result.Failure(AddonMigrationException(failure))
+            }
+        } catch (e: Exception) {
+            return Result.Failure(AddonMigrationException(AddonMigrationResult.Failure.FailedToQueryInstalledAddons(e)))
+        }
+    }
+
+    private suspend fun getInstalledAddons(engine: Engine): List<WebExtension> {
+        return CompletableDeferred<List<WebExtension>>().also { result ->
+            withContext(Dispatchers.Main) {
+                engine.listInstalledWebExtensions(
+                    onSuccess = { result.complete(it) },
+                    onError = { result.completeExceptionally(it) }
+                )
+            }
+        }.await()
+    }
+
+    private suspend fun migrateAddon(engine: Engine, addon: WebExtension): WebExtension {
```

Comment:
Is the `addon` that's returned here actually different from `addon` that's passed-in? Or are you just round-tripping it (in which case, that seems unnecessary)?

Reply:
Yes, the `addons` are different. The returned once have their enabled state updated. We want this so we can check the successfully migrated addons and we'd know if they were enabled/disabled (supported/unsupported). So, I think this makes sense to keep.

Reply:
👍 I thought it may be about the state flip, thank you for confirming.